src/server/dflycmd.cc

```diff
@@ -336,6 +339,56 @@ void DflyCmd::StartStable(CmdArgList args, ConnectionContext* cntx) {
   return rb->SendOk();
 }
 
+void DflyCmd::TakeOver(CmdArgList args, ConnectionContext* cntx) {
+  RedisReplyBuilder* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
+  string_view sync_id_str = ArgS(args, 2);
+  float timeout;
+  if (!absl::SimpleAtof(ArgS(args, 1), &timeout)) {
+    return (*cntx)->SendError(kInvalidIntErr);
+  }
+  if (timeout < 0) {
+    return (*cntx)->SendError("timeout is negative");
+  }
+
+  VLOG(1) << "Got DFLY TAKEOVER " << sync_id_str;
+
+  auto [sync_id, replica_ptr] = GetReplicaInfoOrReply(sync_id_str, rb);
+  if (!sync_id)
+    return;
+
+  unique_lock lk(replica_ptr->mu);
+  if (!CheckReplicaStateOrReply(*replica_ptr, SyncState::STABLE_SYNC, rb))
+    return;
+
+  VLOG(1) << "Takeover initiated, locking down the database.";
+  TransactionGuard tg{cntx->transaction};
+  AggregateStatus status;
+
+  auto cb = [replica_ptr = replica_ptr, timeout, &status](EngineShard* shard) {
+    FlowInfo* flow = &replica_ptr->flows[shard->shard_id()];
+
+    shard->journal()->RecordEntry(0, journal::Op::PING, 0, 0, {}, true);
+
+    absl::Time start = absl::Now();
+    while (flow->last_acked_lsn < shard->journal()->GetLsn()) {
+      if (absl::ToInt64Seconds(absl::Now() - start) > timeout) {
+        VLOG(1) << "Couldn't synchronize with replica for takeover in time.";
+        status = OpStatus::TIMED_OUT;
+        return;
+      }
+      ThisFiber::SleepFor(1ms);
```

Comment:
Did you already plan any error handling? If the replica fails, we will always wait till the end, because cancellation can't run as long as you hold the replica_ptr's mutex.

Is quick error handling important here at all?

Reply:
I added a check for cntx.IsCancelled(), which I think is enough. I don't think it should be interruptible with manual commands